persistence/rocksdb/src/main/java/org/infinispan/persistence/rocksdb/RocksDBStore.java

```diff
@@ -634,4 +516,580 @@ public RocksEntryIterator(RocksIterator it, Predicate<? super K> filter, boolean
+                }
+            }
+        }
+
+        @Override
+        void close() {
+            defaultColumnFamilyHandle.close();
+
+            db.close();
+        }
+
+        protected void reinitAllDatabases() throws IOException, RocksDBException {
+            try {
+                semaphore.acquire(Integer.MAX_VALUE);
+            } catch (InterruptedException e) {
+                throw new PersistenceException("Cannot acquire semaphore", e);
+            }
+            try {
+                if (stopped) {
+                    throw new PersistenceException("RocksDB is stopped");
+                }
+                db.close();
+                expiredDb.close();
+                // Force a GC to ensure that open file handles are released in Windows
+                System.gc();
+                String dataLocation = getQualifiedLocation();
+                Util.recursiveFileRemove(new File(dataLocation));
+                db = open(getQualifiedLocation(), dataDbOptions());
+
+                String expirationLocation = getQualifiedExpiredLocation();
+                Util.recursiveFileRemove(new File(expirationLocation));
+                expiredDb = openDatabase(expirationLocation, expiredDbOptions());
+            } finally {
+                semaphore.release(Integer.MAX_VALUE);
+            }
+        }
+
+        protected RocksIterator wrapIterator(RocksDB db, int segment) {
+            // Some Cache Store tests use clear and in case of the Rocks DB implementation
+            // this clears out internal references and results in throwing exceptions
+            // when getting an iterator. Unfortunately there is no nice way to check that...
+            return db.newIterator(defaultColumnFamilyHandle, new ReadOptions().setFillCache(false));
+        }
+
+        @Override
+        Publisher<K> publishKeys(IntSet segments, Predicate<? super K> filter) {
+            Predicate<? super K> combinedFilter = PersistenceUtil.combinePredicate(segments, keyPartitioner, filter);
+            return publish(-1, it -> Flowable.fromIterable(() -> new RocksKeyIterator(it, combinedFilter)));
+        }
+
+        @Override
+        Publisher<MarshalledEntry<K, V>> publishEntries(IntSet segments, Predicate<? super K> filter, boolean fetchValue,
+              boolean fetchMetadata) {
+            Predicate<? super K> combinedFilter = PersistenceUtil.combinePredicate(segments, keyPartitioner, filter);
+            return publish(-1, it -> Flowable.fromIterable(() -> {
+                // Make sure this is taken when the iterator is created
+                long now = timeService.wallClockTime();
+                return new RocksEntryIterator(it, combinedFilter, fetchValue, fetchMetadata, now);
+            }));
+        }
+
+        @Override
+        void addSegments(IntSet segments) {
+            // Do nothing
+        }
+
+        @Override
+        void removeSegments(IntSet segments) {
+            clear(segments);
```

Comment:
Shouldn't this be a no-op?

Reply:
Nope. This method would be invoked during state transfer when segments are removed from this node. Since the non segmented implementation can't easily remove the segments it has to clear out the data pertaining to those segments.